Add the `OpportunityCard` component

realsimonwu opened the pull request:
I've created the `OpportunityCard` component, as per the [Figma](https://www.figma.com/file/fj6Oxp1IHLN55IZR9V2keV/PuMP-Direct).

The breakpoint for the swap to mobile layout card will probably have to be adjusted later in accordance with how the overall Direct Opportunities page is built.

How it looks at about 1000 pixels wide (about how I expect it to be at full desktop width, because the filtering tags and all that will be beside the cards from the Figma design):
![image](https://user-images.githubusercontent.com/46069730/130326009-8baf374b-a44d-47ee-a485-d1f7d7658cd0.png)

I selected 800 pixels kind of arbitrarily as a breakpoint to switch to the mobile font sizes and card layout: it looks a little odd, so it will have to be adjusted with the Direct Opportunities page once that's built.
![image](https://user-images.githubusercontent.com/46069730/130326019-90cd35f5-083d-43b2-979e-72e4e21fc310.png)

This is what the cards would look like at a more expected mobile resolution.
![image](https://user-images.githubusercontent.com/46069730/130326023-c4ef6eff-f855-4f56-b4b1-6b7e1099e706.png)

realsimonwu committed:
Initial: create the card

realsimonwu committed:
Merge branch 'main' into create-direct-opportunities-card

realsimonwu committed:
Prettier formatting and removing some stuff that I import or define and never use

realsimonwu committed:
Merge branch 'create-direct-opportunities-card' of https://github.com/PuMProfessionals/pump-web-v2 into create-direct-opportunities-card

xsharonhe commented:
> I've created the `OpportunityCard` component, as per the [Figma](https://www.figma.com/file/fj6Oxp1IHLN55IZR9V2keV/PuMP-Direct).
> 
> The breakpoint for the swap to mobile layout card will probably have to be adjusted later in accordance with how the overall Direct Opportunities page is built.
> 
> How it looks at about 1000 pixels wide (about how I expect it to be at full desktop width, because the filtering tags and all that will be beside the cards from the Figma design):
> ![image](https://user-images.githubusercontent.com/46069730/130326009-8baf374b-a44d-47ee-a485-d1f7d7658cd0.png)
> 
> I selected 800 pixels kind of arbitrarily as a breakpoint to switch to the mobile font sizes and card layout: it looks a little odd, so it will have to be adjusted with the Direct Opportunities page once that's built.
> ![image](https://user-images.githubusercontent.com/46069730/130326019-90cd35f5-083d-43b2-979e-72e4e21fc310.png)
> 
> This is what the cards would look like at a more expected mobile resolution.
> ![image](https://user-images.githubusercontent.com/46069730/130326023-c4ef6eff-f855-4f56-b4b1-6b7e1099e706.png)

Could you implement this on the opportunity.jsx page first before I review?

realsimonwu committed:
Rough implementation of `OpportunityCard` on opportunities page

Things are very rough, but it's kind of working!

realsimonwu commented:
Rough implementation of the `OpportunityCard` in the opportunities page.

![image](https://user-images.githubusercontent.com/46069730/130360519-2ae5a855-2091-4f5b-879f-a841d944824d.png)
![image](https://user-images.githubusercontent.com/46069730/130360533-ae61ae53-6182-493a-b5ea-e164bef86ac8.png)
![image](https://user-images.githubusercontent.com/46069730/130360548-a03255ba-2f8e-4a3f-8c4e-adbf2df91db0.png)

Jilly's filtering is working very well!

As of now, there are quite a few things I still need to work out for the cards:
* I need to make the cards clickable, to access each opportunity. As of right now just the opportunity names are clickable links, because I threw this together very quickly.
* I need to put the tags from each opportunity into the cards.
![image](https://user-images.githubusercontent.com/46069730/130360621-2a7cd8fc-7dde-487e-a5fd-63bd71fd2d40.png)
* The cards have padding issues, and text wrapping breaks EVERYTHING.
![image](https://user-images.githubusercontent.com/46069730/130360677-85bf9409-8cbf-44db-b290-a3b4db45ad29.png)
![image](https://user-images.githubusercontent.com/46069730/130360643-a0b47259-ebb2-4144-9610-d2c74d2f92ab.png)

In addition, some considerations for the overall opportunities page:
* Make 'ResultsWrapper' stretch out across the rest of the width of the page at larger widths or not?

Finally, I spotted an issue with the individual opportunities pages: the back arrows on the opportunity pages point to `/opportunities` and not `/resources/opportunities`. 
![image](https://user-images.githubusercontent.com/46069730/130360810-796bda5f-e66d-400e-944b-9928c61248de.png)

xsharonhe commented on components/OpportunityCard.jsx:
```diff
@@ -0,0 +1,146 @@
+  }
+`;
+
+const Logo = styled.img`
```
What happens when you use next's Image? The logos are cutting off and I'd rather have the full logo with some white space around it than how it's being cut off now.

realsimonwu replied:
logos no longer cut off, now they will scale down to fit

xsharonhe reviewed (commented):
I think you can adapt away from Figma and make it on mobile for the image to appear on top for mobile widths because this will break mobile display since the opportunity card is too big.

Generally, looking good

xsharonhe commented:
Going to ask @hirundinidae or @jilly477 for additional reviews on how the front-end should look

hirundinidae commented on components/OpportunityCard.jsx:
```diff
@@ -0,0 +1,146 @@
+  justify-content: space-between;
+`;
+
+const Name = styled.a`
```
Make this an `<h2>` and link the whole card

hirundinidae commented on components/OpportunityCard.jsx:
```diff
@@ -0,0 +1,146 @@
+  padding: 10px;
+  margin: 10px;
+  align-items: center;
+  @media only screen and (max-width: 900px) {
+    height: 100px;
+    padding: 5px;
```
Is there a reason for not using the media util? 
```
${media(
"breakpoint", 
`styles`
)};
```

realsimonwu replied:
media util doesn't interact well with getting stuff from props or `theme` - it's been done this way in other parts of the codebase for similar reasons

hirundinidae replied:
I don't think there's an issue with media util if you're not using the theme/props inside of it (but I know other media queries here do) 
```
${media (
  "tablet", 
  `
    height: 100px; 
    padding: 5px;
  `
)}; 
``` 
^ so that should be fine here

realsimonwu replied:
yeah that's what I meant, I copied the `$media` statement for all of the styles so I'll change everything else over to use the media util if it doesn't need to use the theme or props

hirundinidae commented on pages/resources/opportunities.jsx:
```diff
@@ -124,13 +124,17 @@ export default function Opportunities({ opps, ...props }) {
-                            {postingName}
-                          </a>
-                        </Link>
+												<OpportunityCard
+													name={postingName}
+													link={`/resources/opportunities/${slug}`}
+													organization={orgName}
+													location={city}
+													category="placeholder"
+													postingDate={postedDate}
+													logo={orgImages}
+												/>
```
Wrap this in 
```
<Link href=`/resources/opportunities/${slug}`>
  <a>
    <OpportunityCard ... /> 
  </a> 
</Link>
``` 
to link the card instead of the opportunity title

realsimonwu replied:
See my comment on how this would interact with assistive technologies: I might end up doing this anyways, but it's something I wanted to bring up for consideration just in case it's something we need to address

hirundinidae replied:
Sure, we can probably discuss just linking the blog/opportunity title. This is just if you're having trouble with linking the whole card since you said: 
>  I need to make the cards clickable, to access each opportunity. As of right now just the opportunity names are clickable links, because I threw this together very quickly.

hirundinidae reviewed (commented):
Agree with Sharon that you can move away from the mobile design, either move to 2 rows (logo, then info, more square-like) or omit the logo on mobile and only show opp info (long rectangle). I'm pretty sure the mobile design on figma shows text that is too small to read. 
> I need to make the cards clickable, to access each opportunity. As of right now just the opportunity names are clickable links, because I threw this together very quickly.

You can remove the link from the title and wrap each opportunity card in `<Link>` tags. The blog cards also have hover transition if you want to add the same to these. 
> The cards have padding issues, and text wrapping breaks EVERYTHING.

I think this is caused by the height being set in pixels in the Wrapper. I'd also set the padding/margins with % to increase as the screen width increases (or generally could use more padding around the cards). 
> Make 'ResultsWrapper' stretch out across the rest of the width of the page at larger widths or not?

Make the div stretch the remainder of the page, but horizontally center the opportunity cards within the div? 

Also can't see many of the opp logos.

realsimonwu commented:
> Agree with Sharon that you can move away from the mobile design...

This is the direction I'll probably go in. I won't have much time to get this done for a couple days because my summer course is ending tomorrow and there are some last-minute summatives to polish, but after that I'll be able to work on it.

> You can remove the link from the title and wrap each opportunity card...

I only have one small concern with this and that is that from my research, wrapping cards and other components in the `<a>` tag doesn't interact with screen readers and some other accessibility features very well. From [here](https://css-tricks.com/block-links-the-search-for-a-perfect-solution/#method-4-sprinkle-javascript-on-the-second-method):

> ...the content is not semantic, so assistive technology will announce everything inside the card...

I'm not really sure how big of a consideration this is, because I'm not really sure how good our site is overall in terms of being accessible through screen readers and other assistive technologies. Still, it's something I still think should be taken into consideration.

> Make the div stretch the remainder of the page, but horizontally center the opportunity cards within the div?

This will probably be fine.

> Also can't see many of the opp logos.

Many of the opportunity files still don't have accompanying images to go with them, and are marked as such in their `.mdx` files, so no image renders for those.

![image](https://user-images.githubusercontent.com/46069730/130537712-a6dcfdfe-6260-43e1-b4b3-6c9cd6f7d794.png)

It might be a good idea to replace all of those `(fill in)`'s in the opportunity files with the path to a nice placeholder image. That's something I might end up doing myself.

xsharonhe commented:
> > Agree with Sharon that you can move away from the mobile design...
> 
> This is the direction I'll probably go in. I won't have much time to get this done for a couple days because my summer course is ending tomorrow and there are some last-minute summatives to polish, but after that I'll be able to work on it.
> 
> > You can remove the link from the title and wrap each opportunity card...
> 
> I only have one small concern with this and that is that from my research, wrapping cards and other components in the `<a>` tag doesn't interact with screen readers and some other accessibility features very well. From [here](https://css-tricks.com/block-links-the-search-for-a-perfect-solution/#method-4-sprinkle-javascript-on-the-second-method):
> 
> > ...the content is not semantic, so assistive technology will announce everything inside the card...
> 
> I'm not really sure how big of a consideration this is, because I'm not really sure how good our site is overall in terms of being accessible through screen readers and other assistive technologies. Still, it's something I still think should be taken into consideration.
> 
> > Make the div stretch the remainder of the page, but horizontally center the opportunity cards within the div?
> 
> This will probably be fine.
> 
> > Also can't see many of the opp logos.
> 
> Many of the opportunity files still don't have accompanying images to go with them, and are marked as such in their `.mdx` files, so no image renders for those.
> 
> ![image](https://user-images.githubusercontent.com/46069730/130537712-a6dcfdfe-6260-43e1-b4b3-6c9cd6f7d794.png)
> 
> It might be a good idea to replace all of those `(fill in)`'s in the opportunity files with the path to a nice placeholder image. That's something I might end up doing myself.

Yup, we need this done ASAP it doesn't have to be perfect and you should adapt to the design (it's not definitive just a guide). If you can't finish within two days or so, let me know and I'll do it so we can get testing with everything ASAP.

realsimonwu commented:
Everything for my school course has to be done by 4pm tomorrow, so I know for sure that I'll have time in the evening tomorrow and then the day after to work on this - I'll probably have it done within 2 or 3 days. I'll let you know for sure if anything else comes up

jilly477 reviewed (commented):
I also agree, I personally feel like maybe doing it so that there is no logo on mobile versions may be better to accommodate the different sizes each place's logo will be. 

In terms of the breaking point, I feel like 900 is a little too small as things start looking weird around 950px for me. 
![image](https://user-images.githubusercontent.com/67667948/130559830-915b1b19-96ea-44a9-bcd7-d63fc72ff330.png)

On the blog page, I edited the filtering design so that certain margins and sizing changes at around 1050px so maybe you can try that as a breakpoint. 

Also, I didn't do this for opportunities.jsx yet but there were some new formatting css written for the blog.jsx filtering section. Would you be able to add it to this pull request as well? It's just a few changes to the css at the bottom of opportunities.jsx and you can find exactly what to replace in this commit in a past pull request. The change starts at line 152: https://github.com/PuMProfessionals/pump-web-v2/pull/61/commits/d2c9c9f45d031cac771099b913c8638611df8622

jilly477 reviewed (commented):
I also noticed that when I filter opportunities the length of the cards seem to change to the largest card out of the results. Is there a way to keep all cards the same size at all time? 
Before:
![image](https://user-images.githubusercontent.com/67667948/130561393-d48ff1f0-e660-4d0d-9eb9-4b0938c48152.png)

After:
![image](https://user-images.githubusercontent.com/67667948/130561436-859c8355-2641-4ab0-963d-ff740795ec99.png)

I don't think this happens when it hits the tablet breakpoint, just on the bigger screen sizes

realsimonwu committed:
Style changes to `opportunities` brought over from the blog page

realsimonwu committed:
Make the cards clickable

realsimonwu committed:
Make the `ResultsWrapper` go the rest of the length of the page, and improve the cards

Text now actually has space to wrap, the pictures for opportunities have been reformatted in a way that I like

realsimonwu committed:
Slight formatting changes to the cards, + tags now show up on the cards

realsimonwu committed:
Use the media util

realsimonwu committed:
replace `(fill in)` on opportunity files with a filler image, and also fix a pathname typo

realsimonwu committed:
undoing stuff for merge conflict issues

realsimonwu commented:
Things should be looking much better now. The cards can now have text wrap inside of them without being destroyed, the image location on the cards has changed and looks way cooler now (in my opinion), the `ResultsWrapper` now stretches the rest of the way across the page and the cards fill it completely, the cards are now fully linked, tags are displaying on the cards now, and the style changes that Jilly requested for the filtering section have been added.

What remains to be determined:
* Background color for the images on the cards.
  * Readability of the images (especially those with transparent backgrounds) vs. contrast with the rest of the card.

No background:
![image](https://user-images.githubusercontent.com/46069730/130792092-7ec498c4-431b-4024-9bfa-ff9af9653bde.png)

Darker background for images:
![image](https://user-images.githubusercontent.com/46069730/130792062-77ef9abb-20a6-4a4b-8bee-ff6e1dcb5dc7.png)

Show on the right of the image (I've committed this one for now but we can change it):
![image](https://user-images.githubusercontent.com/46069730/130792170-11b59d67-ea0a-47b6-88cf-1403cdb20630.png)

realsimonwu committed:
Import `media` util because I forgot to before

realsimonwu commented:
A decision has been reached! The shadow to the right of the image on each card will stay.

xsharonhe reviewed (commented):
Much better but I'm a little concerned about how it will look on really small phones as it starts breaking less than 350. I'll let Diana be the one to approve

hirundinidae reviewed (commented):
Looks better, just running into problems on some cards at 360px as Sharon mentioned. (cards with longer opportunity names)  
![screenshot_20210827_194736](https://user-images.githubusercontent.com/65516245/131199254-1d545388-3637-401b-a120-af557fe47712.png)

hirundinidae commented on components/OpportunityCard.jsx:
```diff
@@ -0,0 +1,168 @@
+  display: flex;
+  flex-direction: row;
+	width: calc(100%-40px);
+  height: 190px;
+  padding: 0;
+  margin: 10px;
```
The margin isn't showing up at smaller widths (card doesn't grow as the content gets closer to the edge), I think because the height is forcing the card to only be so tall.

realsimonwu committed:
Hide text on overflow

xsharonhe commented:
@noisya55616 I'm merging this now - You can make your changes in another MR.

realsimonwu commented:
Just to let you know - the lastest commit I made will hide the information if long opportunity names make the card overflow, so that nothing spills out of the cards

realsimonwu commented:
It has already been included in the merge you just did
